store/pogreb/pogreb.go

```diff
@@ -35,52 +39,130 @@ func New(dir string) (*pStorage, error) {
+	entryDir := path.Join(dir, "entries")
+	es, err := pogreb.Open(entryDir, &opts)
+	if err != nil {
+		return nil, err
+	}
+	return &pStorage{dir: dir, store: s, entStore: es}, nil
 }
 
 func (s *pStorage) Get(c cid.Cid) ([]entry.Value, bool, error) {
-	return s.get(c.Bytes())
+	out := []entry.Value{}
+	ks, found, err := s.get(c.Bytes())
+	if err != nil {
+		return nil, false, err
+	}
+	if !found {
+		return nil, found, nil
+	}
+	// If keys found, get entries from entryStore
+	for i := range ks {
+		e, found, err := s.getEntry(ks[i])
+		if err != nil {
+			return nil, false, err
+		}
+		if !found {
+			return nil, false, errors.New("one of the entries couldn't be found in entryStore")
+		}
+		out = append(out, e.Value)
+	}
+
+	return out, true, nil
 }
 
-func (s *pStorage) get(k []byte) ([]entry.Value, bool, error) {
+func (s *pStorage) get(k []byte) (store.CidEntry, bool, error) {
 	value, err := s.store.Get(k)
 	if err != nil {
 		return nil, false, err
 	}
+	if value == nil {
+		return [][]byte{}, false, nil
+	}
+
+	return store.SplitKs(value), true, nil
+}
+
+func (s *pStorage) getEntry(k []byte) (*store.Entry, bool, error) {
+	value, err := s.entStore.Get(k)
+	if err != nil {
+		return nil, false, err
+	}
 	if value == nil {
 		return nil, false, nil
 	}
 
-	out, err := entry.Unmarshal(value)
+	out, err := store.Unmarshal(value)
 	if err != nil {
 		return nil, false, err
 	}
 	return out, true, nil
-
 }
 
 func (s *pStorage) Put(c cid.Cid, entry entry.Value) (bool, error) {
 	return s.put(c.Bytes(), entry)
 }
 
 func (s *pStorage) put(k []byte, in entry.Value) (bool, error) {
+	entK, err := store.EntryKey(in)
+	if err != nil {
+		return false, err
+	}
 	old, found, err := s.get(k)
 	if err != nil {
 		return false, err
 	}
 	// If found it means there is already a value there.
 	// Check if we are trying to put a duplicate entry
-	if found && duplicateEntry(in, old) {
+	if found && store.DuplicateEntry(entK, old) {
 		return false, nil
 	}
 
-	li := append(old, in)
-	b, err := entry.Marshal(li)
+	// If no duplicate put the entry.
+	ok, err := s.putEntry(entK, in)
 	if err != nil {
 		return false, err
 	}
+	// Ok needs to be true, RefC needs to be increased successfully
+	if !ok {
+		return false, errors.New("failed to put entry in entryStore")
+	}
+	li := append(old, entK)
 
-	err = s.store.Put(k, b)
+	err = s.store.Put(k, store.JoinKs(li))
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *pStorage) putEntry(k []byte, in entry.Value) (bool, error) {
+	old, found, err := s.getEntry(k)
+	if err != nil {
+		return false, err
+	}
+	// Found in store. Increase RefC and update in store
+	if found {
+		old.RefC++
+		b, err := store.Marshal(old)
+		if err != nil {
+			return false, err
+		}
+		s.entStore.Put(k, b)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+
+	// If not found the entry is new and needs to be fully put.
+	e := &store.Entry{Value: in, RefC: 1}
```

Comment:
Race condition?  If two concurrent calls both find that the entry is not found, and the both set the reference count to 1, but really it should be 2.